Feature/5 releases tab

NightOfNights opened the pull request.

NightOfNights committed:
chore: implement function for getting language codes from releases #5

NightOfNights committed:
Merge branch 'develop' into feature/5-releases-tab

NightOfNights committed:
chore: change default navigation link #5

NightOfNights committed:
chore: make release date non-nullable  #5

NightOfNights committed:
feature: implement date and country services #5

NightOfNights committed:
feature: implement base layout for releases #5

NightOfNights committed:
Merge remote-tracking branch 'origin/feature/1-implement-layout' into feature/5-releases-tab

NightOfNights committed:
Merge remote-tracking branch 'origin/feature/1-implement-layout' into feature/5-releases-tab

NightOfNights committed:
fix: return release model to how it was before #5

NightOfNights committed:
fix: resolve sidebar key error #5

NightOfNights committed:
feature: improve fetch releases query #5
add new countries

NightOfNights committed:
fix: resolve overview page error #5

NightOfNights committed:
fix: check tba case for released date #5

NightOfNights committed:
Merge branch 'feature/5-releases-tab' of github.com:vndbre/vndbre into feature/5-releases-tab

NightOfNights committed:
feature: add platform service #5
rename country service to language service, improve release mapper

NightOfNights committed:
fix: resolve problem with default layout #5

NightOfNights committed:
feature: add all remaining platforms #5
Add default icon for platforms

NightOfNights committed:
feature: add all remaining languages #5
Add default icon for language

NightOfNights committed:
Merge branch 'develop' into feature/5-releases-tab

NightOfNights committed:
fix: resolve active link bug on empty route #5

NightOfNights committed:
feature: add release status #5
wrap component with content wrapper

NightOfNights committed:
refactor: format releases page component #5

NightOfNights committed:
fix: improve docs #5

NightOfNights committed:
refactor: move out release enums #5
update eslint rule for enums to ingore export duplicate error

NightOfNights committed:
feature: add release icons and link to website #5

NightOfNights committed:
fix: make error in content wrapper optional #5

NightOfNights committed:
refactor: move out enums folder from models #5

NightOfNights committed:
chore: add utility function for enums #5

NightOfNights committed:
fix: make getIsFreewareIcon method return right icon
#5

NightOfNights committed:
Merge branch 'develop' into feature/5-releases-tab

NightOfNights committed:
Merge branch 'feature/1-implement-layout' into feature/5-releases-tab

NightOfNights committed:
fix: resolve error after merge #5

NightOfNights committed:
refactor: move out some things to another files #5
improve namings, add toReadable method to ReleaseType enum

moetazaneta commented on src/features/VisualNovel/pages/ReleasesPage/ReleasesPage.tsx:
```diff
@@ -1,24 +1,168 @@
+        {} as ReleaseGroups,
+      );
+      return { ...accumulatedReleases, ...release };
+    }, {} as ReleaseGroups);
```
It's better to use generic instead of type assertion, like so `.reduce<ReleaseGroups>(`

moetazaneta commented on src/api/mappers/releaseMapper.ts:
```diff
@@ -32,6 +33,20 @@ const releaseProducerFromDto = (dto: ReleaseProducerDto): ReleaseProducer => ({
+ * @param minAge Minimum age.
+ */
+const mapMinAgeToRating = (minAge: number | null): string => {
+  if (minAge !== null) {
+    if (minAge === 0) {
+      return 'All ages';
+    }
+    return `${minAge}+`;
+  }
+  return '';
```
I think it's better better to avoid unnecessary nesting 😇
```suggestion
  if (minAge === null) {
    return '';
  }
  if (minAge === 0) {
    return 'All ages';
  }
  return `${minAge}+`;
```

moetazaneta commented on src/api/services/languageService.ts:
```diff
@@ -0,0 +1,139 @@
+  Korean = 'ko',
+  Vietnamese = 'vi',
+  Chinese = 'zh',
+  Spanish = 'es',
+  Turkish = 'tr',
+  German = 'de',
+  PortugueseBrazil = 'pt-br',
+  Polish = 'pl',
+  Hungarian = 'hu',
+  Czech = 'cs',
+  Ukrainian = 'uk',
+  Catalan = 'ca',
+  Arabic = 'ar',
+  Bulgarian = 'bg',
+  Swedish = 'sv',
+  Finnish = 'fi',
+  Romanian = 'ro',
+  Tagalog = 'ta',
+  Macedonian = 'mk',
+  Slovene = 'sl',
+  Hebrew = 'he',
+  Indonesian = 'id',
+  Danish = 'da',
+  Latvian = 'lv',
+  Lithuanian = 'lt',
+  Dutch = 'nl',
+  Esperanto = 'eo',
+  Greek = 'el',
+  Hindi = 'hi',
+  Thai = 'th',
+  Urdu = 'ur',
+  Persian = 'fa',
+  Irish = 'ga',
+  Latin = 'la',
+  Portuguese = 'pt-pt',
+  Malay = 'ms',
+  Norwegian = 'no',
+  Slovak = 'sk',
+  ScottishGaelic = 'gd',
+}
```
Is it all possible languages?
If no, I think we need some default/other field.

moetazaneta commented on src/api/services/languageService.ts:
```diff
@@ -0,0 +1,139 @@
+   * Converts a certain language to readable equivalent.
+   * @param value Language.
+   */
+  export const toReadable = (value: Language): string => MAP_LANGUAGE_INFO[value]?.name ?? '';
```
Isn't it impossible for `MAP_LANGUAGE_INFO[value]?.name` to be nullish?
So I think optional chaining and nullish operator are unnecessary.

moetazaneta commented on src/api/services/releaseService.ts:
```diff
@@ -2,19 +2,190 @@ import { http } from '../index';
+
+    while (hasToFetchMore) {
+    // eslint-disable-next-line no-await-in-loop
+      const releasesData = await fetch(currentPage);
```
why `fetch`?

NightOfNights replied:
It's local `fetch` function)

moetazaneta replied:
x)

moetazaneta commented on src/api/services/releaseService.ts:
```diff
@@ -2,19 +2,190 @@ import { http } from '../index';
+      }
+    }
+
+    return releasesChunks.map(releasesChunk => releasesChunk.map(dto => releaseFromDto(dto))).flat();
```
I suggest `flatMap`.

moetazaneta commented on src/api/services/releaseService.ts:
```diff
@@ -2,19 +2,190 @@ import { http } from '../index';
+   */
+  export const getReleaseStatusIcon = (releaseType: ReleaseType): ReleaseIcon => {
+    const label = ReleaseType.toReadable(releaseType);
+    const icon = (() => {
+      switch (releaseType) {
+        case ReleaseType.Trial:
+          return 'carbon:circle-dash';
+        case ReleaseType.Partial:
+          return 'carbon:incomplete';
+        default:
+          return 'carbon:circle-solid';
+      }
+    })();
```
I think it's a bit cleaner.
```suggestion
  const icon = {
    [ReleaseType.Trial]: 'carbon:circle-dash',
    [ReleaseType.Partial]: 'carbon:incomplete',
    [ReleaseType.Complete]: 'carbon:circle-solid',
  }[releaseType];
```

moetazaneta commented on src/enums/releaseAnimationType.ts:
```diff
@@ -0,0 +1,27 @@
+   * Converts a certain release animation type to readable equivalent.
+   * @param value Type of animation of release.
+   */
+  export function toReadable(value: ReleaseAnimationType): string {
```
Why it's have different code style from other `toReadable`'s (not arrow function, no checks)?

moetazaneta commented:
Add `padding-bottom: var(--chakra-sizes-8);` to `.content` in `DefaultLayout.module.css` please, so it look better.

moetazaneta commented on src/features/VisualNovel/pages/ReleasesPage/ReleasesPage.tsx:
```diff
@@ -1,24 +1,168 @@
+              <Text>{release.releasedISODate}</Text>
+              <Box display="flex" gridGap={1}>
+                {getReleaseStatusElement(release.type)}
+                <Text fontWeight="bold">{release.title}</Text>
```
Should be `semibold`.

moetazaneta commented on src/features/VisualNovel/pages/ReleasesPage/ReleasesPage.tsx:
```diff
@@ -1,24 +1,168 @@
+                ))}
+              </Box>
+              {release.website && (
+                <Link
+                  borderBottomColor="transparent"
```
```suggestion
                <Link
                  variant="no-underline"
```

NightOfNights committed:
refactor: improve code after code review #5

NightOfNights committed:
refactor: remove unused method #5

Moltenship commented on src/features/VisualNovel/pages/ReleasesPage/ReleasesPage.tsx:
```diff
@@ -1,24 +1,168 @@
+                {getReleaseStatusElement(release.type)}
+                <Text fontWeight="semibold">{release.title}</Text>
+              </Box>
+              <Box display="flex" gridGap={2}>
```
You can use `Stack` or `HStack` element from `chakra-ui`. These components comes with flex by default.

NightOfNights committed:
refactor: update models, remove type assertions #5

NightOfNights committed:
refactor: use hstack and grid instead of box #5

moetazaneta commented on src/api/services/languageService.ts:
```diff
@@ -0,0 +1,131 @@
+   * Converts a certain language to readable equivalent.
+   * @param value Language.
+   */
+  export const toReadable = (value: Language): string => MAP_LANGUAGE_INFO[value]?.name ?? MAP_LANGUAGE_INFO[Language.OtherLanguage];
```
Shouldn't it be `MAP_LANGUAGE_INFO[Language.OtherLanguage].name`?

NightOfNights committed:
refactor: improve release dto #5
create method for language and platform services that convert string to needed type, then use it in releaseMapper to map string value from dto to needed type in release model
